File: internal/models/project.go
package models

import "time"

type ProjectConfig struct {
	ID          string    `json:"id" db:"id"`
	ServerID    string    `json:"serverId,omitempty" db:"server_id"` // Node where this project runs
	Name        string    `json:"name" db:"name"`
	Description string    `json:"description,omitempty" db:"description"`
	CreatedAt   time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt   time.Time `json:"updatedAt" db:"updated_at"`
}

type SSLCertStatus string

const (
	SSLCertStatusPending SSLCertStatus = "pending"
	SSLCertStatusIssued  SSLCertStatus = "issued"
	SSLCertStatusFailed  SSLCertStatus = "failed"
)

type MemberPermission string

const (
	MemberPermissionOwner  MemberPermission = "owner"
	MemberPermissionAdmin  MemberPermission = "admin"
	MemberPermissionMember MemberPermission = "member"
)

type MemberStatus string

const (
	MemberStatusPending  MemberStatus = "pending"
	MemberStatusAccepted MemberStatus = "accepted"
)

type DomainConfig struct {
	ID            string        `json:"id" db:"id"`
	ServiceID     string        `json:"serviceId" db:"service_id"`
	DomainName    string        `json:"domainName" db:"domain_name"`
	RedirectTo    string        `json:"redirectTo,omitempty" db:"redirect_to"`
	SSLCertStatus SSLCertStatus `json:"sslCertStatus" db:"ssl_cert_status"`
	PathPrefix    string        `json:"pathPrefix" db:"path_prefix"`
	CreatedAt     time.Time     `json:"createdAt" db:"created_at"`
	UpdatedAt     time.Time     `json:"updatedAt" db:"updated_at"`
}

type ServerlessFunctionCode struct {
	ID          string    `json:"id" db:"id"`
	ServiceID   string    `json:"serviceId" db:"service_id"`
	Runtime     string    `json:"runtime" db:"runtime"`
	CodeContent string    `json:"codeContent" db:"code_content"`
	CreatedAt   time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt   time.Time `json:"updatedAt" db:"updated_at"`
}

type EnvironmentConfig struct {
	ID        string    `json:"id" db:"id"`
	ProjectID string    `json:"projectId" db:"project_id"`
	Name      string    `json:"name" db:"name"`
	IsDefault bool      `json:"isDefault" db:"is_default"`
	CreatedAt time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt time.Time `json:"updatedAt" db:"updated_at"`
}

type CanvasSummary struct {
	ID                 string             `json:"id"`
	Name               string             `json:"name"`
	Description        string             `json:"description,omitempty"`
	CreatedAt          time.Time          `json:"createdAt"`
	UpdatedAt          time.Time          `json:"updatedAt"`
	EnvironmentsCount  int                `json:"environmentsCount"`
	AppsCount          int                `json:"appsCount"`
	DatabasesCount     int                `json:"databasesCount"`
	OnlineServices     int                `json:"onlineServices"`
	TotalServices      int                `json:"totalServices"`
	ServiceIcons       []string           `json:"serviceIcons"`
	DefaultEnvironment *EnvironmentConfig `json:"defaultEnvironment,omitempty"`
}

type EnvironmentCanvas struct {
	Environment *EnvironmentConfig `json:"environment"`
	Apps        []*AppService      `json:"apps"`
	Databases   []*Database        `json:"databases"`
}

type CreateProjectRequest struct {
	ID                 string `json:"id"`
	Name               string `json:"name"`
	Description        string `json:"description,omitempty"`
	RepositoryURL      string `json:"repositoryUrl,omitempty"`
	RepositoryURLSnake string `json:"repository_url,omitempty"`
	Branch             string `json:"branch,omitempty"`
	InternalPort       int    `json:"internalPort,omitempty"`
	InternalPortSnake  int    `json:"internal_port,omitempty"`
	Domain             string `json:"domain,omitempty"`
}

type (
	SetEnvVarsRequest map[string]string
	VarsRequest       map[string]string
)
type ProjectToken struct {
	ID            string     `json:"id" db:"id"`
	ProjectID     string     `json:"projectId" db:"project_id"`
	EnvironmentID string     `json:"environmentId" db:"environment_id"`
	Name          string     `json:"name" db:"name"`
	TokenPrefix   string     `json:"tokenPrefix" db:"token_prefix"`
	Scopes        []string   `json:"scopes" db:"-"`
	IPAllowlist   []string   `json:"ipAllowlist" db:"-"`
	ExpiresAt     *time.Time `json:"expiresAt,omitempty" db:"expires_at"`
	CreatedAt     time.Time  `json:"createdAt" db:"created_at"`
}

type ProjectMember struct {
	ID         string           `json:"id" db:"id"`
	ProjectID  string           `json:"projectId" db:"project_id"`
	UserID     string           `json:"userId,omitempty" db:"user_id"`
	Email      string           `json:"email" db:"email"`
	Permission MemberPermission `json:"permission" db:"permission"`
	Status     MemberStatus     `json:"status" db:"status"`
	InvitedAt  time.Time        `json:"invitedAt" db:"invited_at"`
	AcceptedAt time.Time        `json:"acceptedAt,omitempty" db:"accepted_at"`
}

type CreateTokenRequest struct {
	Name          string     `json:"name"`
	EnvironmentID string     `json:"environmentId"`
	Scopes        []string   `json:"scopes"`
	IPAllowlist   []string   `json:"ipAllowlist,omitempty"`
	ExpiresAt     *time.Time `json:"expiresAt,omitempty"`
}

type AddMemberRequest struct {
	Email      string           `json:"email"`
	Permission MemberPermission `json:"permission"`
}

File: dashboard/src/interfaces/project.ts
import type { BaseResponse, PaginatedData } from './base';
import type { Database } from './database';
import type { AppService } from './deployment';
import type { Storage } from './storage';

export type MemberPermission = 'admin' | 'member' | 'viewer';
export type MemberStatus = 'pending' | 'accepted';
export type SSLCertStatus = 'pending' | 'issued' | 'failed';

export interface ProjectConfig {
  id: string;
  name: string;
  description?: string;
  createdAt: string;
  updatedAt: string;
}

export interface EnvironmentConfig {
  id: string;
  projectId: string;
  name: string;
  isDefault: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface DomainConfig {
  id: string;
  projectId: string;
  domainName: string;
  redirectTo?: string;
  sslCertStatus: SSLCertStatus;
  pathPrefix: string;
  createdAt: string;
  updatedAt: string;
}

export interface ServerlessFunctionCode {
  id: string;
  serviceId: string;
  runtime: string;
  codeContent: string;
  createdAt: string;
  updatedAt: string;
}

export interface CanvasSummary {
  id: string;
  name: string;
  description?: string;
  environmentsCount: number;
  appsCount: number;
  databasesCount: number;
  storageCount: number;
  onlineServices: number;
  totalServices: number;
  serviceIcons: string[];
  defaultEnvironment?: EnvironmentConfig;
  createdAt: string;
  updatedAt: string;
}

export interface EnvironmentCanvas {
  environment: EnvironmentConfig;
  apps: AppService[];
  databases: Database[];
  storage: Storage[];
}

export interface CreateProjectRequest {
  name: string;
  description?: string;
  repositoryUrl?: string;
  branch?: string;
  internalPort?: number;
  domain?: string;
}

export interface ProjectToken {
  id: string;
  projectId: string;
  environmentId: string;
  name: string;
  tokenPrefix: string;
  scopes: string[];
  ipAllowlist: string[];
  expiresAt?: string;
  createdAt: string;
}

export interface ProjectMember {
  id: string;
  projectId: string;
  userId?: string;
  email: string;
  permission: MemberPermission;
  status: MemberStatus;
  invitedAt: string;
  acceptedAt?: string;
}

export interface CreateWebhookRequest {
  url: string;
  eventTypes: string[];
  includePrEnvironments: boolean;
}

export interface CreateTokenRequest {
  name: string;
  environmentId: string;
  scopes: string[];
  ipAllowlist?: string[];
  expiresAt?: string;
}

export interface AddMemberRequest {
  email: string;
  permission: MemberPermission;
}

export type ListProjectsResponse = BaseResponse<PaginatedData<ProjectConfig>>;
export type GetProjectResponse = BaseResponse<ProjectConfig>;
export type CreateProjectResponse = BaseResponse<ProjectConfig>;
export type ListEnvironmentsResponse = BaseResponse<EnvironmentConfig[]>;
export type CreateEnvironmentResponse = BaseResponse<EnvironmentConfig>;
export type GetCanvasResponse = BaseResponse<CanvasSummary[]>;
export type GetEnvironmentCanvasResponse = BaseResponse<EnvironmentCanvas>;
export type ListDomainsResponse = BaseResponse<DomainConfig[]>;
export type CreateDomainResponse = BaseResponse<DomainConfig>;
export type ListMembersResponse = BaseResponse<ProjectMember[]>;
export type ListTokensResponse = BaseResponse<ProjectToken[]>;
